GitUI/BranchTreePanel/RepoObjectsTree.Nodes.Submodules.cs

```diff
@@ -0,0 +1,459 @@
+                //
+                // C:/code/gitextensions2/Externals/conemu-inside
+                // C:/code/gitextensions2/Externals/Git.hub
+                // C:/code/gitextensions2/Externals/ICSharpCode.TextEditor
+                // C:/code/gitextensions2/Externals/ICSharpCode.TextEditor/gitextensions
+                // C:/code/gitextensions2/Externals/ICSharpCode.TextEditor/gitextensions/Externals/conemu-inside
+                // C:/code/gitextensions2/Externals/ICSharpCode.TextEditor/gitextensions/Externals/Git.hub
+                // C:/code/gitextensions2/Externals/ICSharpCode.TextEditor/gitextensions/Externals/ICSharpCode.TextEditor
+                // C:/code/gitextensions2/Externals/ICSharpCode.TextEditor/gitextensions/Externals/NBug
+                // C:/code/gitextensions2/Externals/ICSharpCode.TextEditor/gitextensions/GitExtensionsDoc
+                // C:/code/gitextensions2/Externals/NBug
+                // C:/code/gitextensions2/GitExtensionsDoc
+                //
+                // What we want to do is first remove the topModule portion, "C:/code/gitextensions2/", and
+                // then build our tree by breaking up each path into parts, separated by '/'.
+                //
+                // Note that when we break up the paths, some parts are just directories, the others are submodule nodes:
+                //
+                // Externals / ICSharpCode.TextEditor / gitextensions / Externals / Git.hub
+                //  folder          submodule             submodule      folder     submodule
+                //
+                // Input 'nodes' is an array of SubmoduleNodes for all the submodules; now we need to create SubmoduleFolderNodes
+                // and insert everything into a tree.
+
+                var topModule = threadModule.GetTopModule();
+
+                // Build a mapping of top-module-relative path to node
+                var pathToNodes = new Dictionary<string, Node>();
+
+                // Add existing SubmoduleNodes
+                foreach (var node in submoduleNodes)
+                {
+                    pathToNodes[GetNodeRelativePath(topModule, node)] = node;
+                }
+
+                // Create and add missing SubmoduleFolderNodes
+                foreach (var node in submoduleNodes)
+                {
+                    var parts = GetNodeRelativePath(topModule, node).Split('/');
+                    for (int i = 0; i < parts.Length - 1; ++i)
+                    {
+                        var path = string.Join("/", parts.Take(i + 1));
+                        if (!pathToNodes.ContainsKey(path))
+                        {
+                            pathToNodes[path] = new SubmoduleFolderNode(this, parts[i]);
+                        }
+                    }
+                }
+
+                // Now build the tree
+                var rootNode = new DummyNode();
+                var nodesInTree = new List<Node>();
```

Comment:
I would use a Dictionary here. There is a loop inside of another loop and after that linear search in nodesInTree.

Reply:
Pushed new commit. Used a HashSet instead of List. I believe the List was a vestige of when I supported three different ways to display the tree. No longer necessary.

Reply:
Thanks for the patience, good job.

Reply:
Thanks for your feedback :)